apiserver/facades/client/modelmanager/modelmanager.go

```diff
@@ -358,49 +386,71 @@ func (m *ModelManagerAPI) CreateModel(args params.ModelCreateArgs) (params.Model
 		return result, errors.Trace(err)
 	}
 
-	newConfig, err := m.newModelConfig(cloudSpec, args, controllerModel)
-	if err != nil {
-		return result, errors.Annotate(err, "failed to create config")
-	}
+	var model common.Model
+	var st common.ModelManagerBackend
 
-	// Create the Environ.
-	env, err := environs.New(environs.OpenParams{
-		Cloud:  cloudSpec,
-		Config: newConfig,
-	})
-	if err != nil {
-		return result, errors.Annotate(err, "failed to open environ")
-	}
-	if err := env.Create(environs.CreateParams{
-		ControllerUUID: controllerCfg.ControllerUUID(),
-	}); err != nil {
-		return result, errors.Annotate(err, "failed to create environ")
-	}
-	storageProviderRegistry := stateenvirons.NewStorageProviderRegistry(env)
-
-	// NOTE: check the agent-version of the config, and if it is > the current
-	// version, it is not supported, also check existing tools, and if we don't
-	// have tools for that version, also die.
-	model, st, err := m.state.NewModel(state.ModelArgs{
-		Type:            state.ModelTypeIAAS,
-		CloudName:       cloudTag.Id(),
-		CloudRegion:     cloudRegionName,
-		CloudCredential: cloudCredentialTag,
-		Config:          newConfig,
-		Owner:           ownerTag,
-		StorageProviderRegistry: storageProviderRegistry,
-		EnvironVersion:          env.Provider().Version(),
-	})
-	if err != nil {
-		return result, errors.Annotate(err, "failed to create new model")
-	}
-	defer st.Close()
+	if jujucloud.CloudIsCAAS(cloud) {
+		newConfig, err := m.newCAASModelConfig(cloudSpec, args)
+		if err != nil {
+			return result, errors.Annotate(err, "failed to create config")
+		}
 
-	if err = st.ReloadSpaces(env); err != nil {
-		if errors.IsNotSupported(err) {
-			logger.Debugf("Not performing spaces load on a non-networking environment")
-		} else {
-			return result, errors.Annotate(err, "Failed to perform spaces discovery")
+		model, st, err = m.state.NewModel(state.ModelArgs{
+			Type:            state.ModelTypeCAAS,
+			CloudName:       cloudTag.Id(),
+			CloudCredential: cloudCredentialTag,
+			Config:          newConfig,
+			Owner:           ownerTag,
+		})
+		if err != nil {
+			return result, errors.Annotate(err, "failed to create new model")
```

Comment:
Would it be worth possibly adding methods `newCAASModel` and `newIAASModel` to make the code here more bite-sized. Not at all necessary of course.